driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -337,30 +289,49 @@ private PoolingOptions options() {
       }
     }
 
-    if (connections[shardId].isEmpty()) {
-      if (host.convictionPolicy.canReconnectNow()) {
-        if (connectionsPerShard == 0) {
-          maybeSpawnNewConnection(shardId);
-        } else if (scheduledForCreation[shardId].compareAndSet(0, connectionsPerShard)) {
-          for (int i = 0; i < connectionsPerShard; i++) {
-            // We don't respect MAX_SIMULTANEOUS_CREATION here because it's  only to
-            // protect against creating connection in excess of core too quickly
-            manager.blockingExecutor().submit(new ConnectionTask(shardId));
-          }
-        }
-        return enqueue(timeout, unit, maxQueueSize, shardId);
-      }
+    int index = 0;
+    int shId = shardId;
+    do {
+    	synchronized(connections[shId]) {
+    		if ((connections[shId].isEmpty()) && (index < connections.length)) {
+    	    	shId = index++;
+    		} else {
+    			break;
+    		}
+    	}
+    } while(true);
```

Comment:
Why not while (connections[shId].isEmpty() && index < connections.length) { ... } ?
Even better:
```
if (connections[shId].isEmpty()) {
  do {
    shId = (shId + 1) % connections.length;
  } while (connections[shId].isEmpty() && shId != shardId);
}
```
to prevent overloading connections for shard 0.

Reply:
As for the first suggestion (while...) it's not possible because of the synchronize statement. Regarding the later, that's a good idea, thanks.